### PR TITLE
Ulyssess offload blog

### PR DESCRIPTION
**Pull request by samadejacobs**
Ulysses-Offload (FPDT) blog, please see corresponding tutorial page at [link](https://github.com/microsoft/DeepSpeed/pull/6813).

### CODE REVIEW

File: blogs/ulysses-offload/README.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+## Introduction
+
+In the rapidly evolving field of generative AI and scientific ML, the
+ability to train large (language) models with ultra-long context
+capabilities is becoming increasingly important. These models are
+essential for a variety of complex tasks, ranging from understanding
+lengthy documents to image and video generation to processing extensive
+sequences in computational biology. However, training such models
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:35]: The use of `ranging` makes for an awkward reading. I think `such as` is more appropriate, as an example:

```such as understanding lengthy documents, generating images and videos, and processing extensive sequences in computational biology```

File: blogs/ulysses-offload/README.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+efficiently poses significant challenges due to the enormous GPU
+resources and memory required.
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:37]: Could this be simplified to `... the enormous GPU memory required.

File: blogs/ulysses-offload/README.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+Building on our previous project, DeepSpeed Ulysses, which focuses on
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:39]:
```suggestion
Building on DeepSpeed Ulysses, our previous project, which focuses on
```

File: blogs/ulysses-offload/README.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+system optimizations for training extremely long sequence transformer
+models, we are excited to present the Fully Pipelined Distributed
+Transformer (FPDT), also known as Ulysses-Offload, in this release. FPDT
+is an innovative, resource-efficient technique that offers comparable
+benefits to DeepSpeed Ulysses and other previous long-context
+optimization methods, but with a modest hardware budget. FPDT makes
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:45]: Let's make the comparison explicit by saying `lower` instead of `modest`

```suggestion
optimization methods, but with a lower hardware budget. FPDT makes
```

File: blogs/ulysses-offload/README.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+ultra long-context large language models (LLM) training and finetuning
+accessible to everyone, regardless of GPU resources. FPDT enables
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:47]: Minor:  `regardless` seems too strong. 

```suggestion
accessible to everyone, including those with limited GPU resources. FPDT enables
```

File: blogs/ulysses-offload/README.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+and subsequent sections provide additional details on the design and 
+usability of FPDT, followed by experimental results.
+
+## Key Innovations
+
+### 1. Fully Pipelined Distributed Transformer (FPDT)
+
+The core innovation of our work is the Fully Pipelined Distributed
+Transformer (FPDT). This approach leverages a pipelined sequence
+chunking, which allows for the training of LLMs with sequence lengths up
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:61]: I find it awkward to read this phrase: `a pipelined sequence chunking`. Is there a way to simplify?

File: blogs/ulysses-offload/README.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+One of the critical aspects of our approach is the comprehensive
+analysis and optimization of the memory footprint during LLM training.
+We target the reduction of redundant intermediate buffers in both the
+forward and backward passes of the training process. By optimizing the
+use of GPU and host CPU memory, we can train larger models with longer
+sequences without running into GPU memory limitations. This optimization
+is crucial for enabling the training of ultra-long context models on a
+limited number of GPUs. It is worth noting that FPDT memory optimization
+is orthogonal and complementary to model- parameter-focused memory
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:79]: It would be useful to say why this is orthogonal with something like FPDT optimizes memory footprint of sequence/activations while ZeRO/FSDP optimizes memory footprint of model parameters.

File: blogs/ulysses-offload/README.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+Figure 3 gives an example of how to perform the computation of chunk
+*T<sub>m</sub>*. After the alltoall collective communication,
+*GPU<sub>j</sub>* receives
+$\widehat{q}m,\ \widehat{k}m,\ and\ \widehat{v}m$*.* We then fetch the
+previous sequence chunk by chunk from the host memory to
+GPU<sub>j</sub>, and perform online attention with the current
+$\widehat{q}m$ and update the output chunk accordingly. Note that, in a
+strict manner, at any given time, only one set of chunks
+$\widehat{k}i,\ and\ \widehat{v}i\ $is placed on GPU's HBM, reducing the
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:145]: This is not rendering correctly
image

File: blogs/ulysses-offload/README.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+Figure 2: FPDT core design
+
+### 
+
+### Pipelining and Scheduling
+
+FPDT employs a pipelined sequence chunking design to manage the memory
+and computational load efficiently. In traditional Transformer model,
+input QKV tensor can be denoted *\[B, S, H, D\]*, where *B* is batch
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:114]: I don't get the benefit of representing input sequence as quad-tuple with model hidden dimension split into `H` and `D`. I understand that this split matters for Ulyssess algorithm, but if we are claiming generality to other SP algorithms then our descriptions should remain generic.

File: blogs/ulysses-offload/README.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+size, *S* is sequence length, *H* is number of heads and *D* is hidden
+dimension per head. With sequence parallelism such as DeepSpeed Ulysses,
+input tensor is partitioned along sequence dimension across sequence
+parallel group P, that is *\[B, S/P, H,D\]* prior to alltoall collective
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:118]:
```suggestion
parallel group P, that is *\[B, S/P, H, D\]* prior to alltoall collective
```

File: blogs/ulysses-offload/README.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+communication. The alltoall communication gathers partitioned tensor
+along sequence dimension and scatter them along head dimension essential
+transforming tensor from *\[B, S/P, H, D\]* to *\[B,S, H/P, D\]*. In our
```
Comment (tjruwase) [inline blogs/ulysses-offload/README.md:121]:
```suggestion
transforming tensor from *\[B, S/P, H, D\]* to *\[B, S, H/P, D\]*. In our
```

File: blogs/ulysses-offload/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (samadejacobs) [reply blogs/ulysses-offload/README.md:114]: QKV tensor => q, k, v (separate) tensors. Ulysses assumes input sequence/hidden state projected so q,k,v!

[the remark this answers is not in the document]
Comment (tjruwase) [reply blogs/ulysses-offload/README.md:114]: The confusion is the overload of `input` since the general conversation is about sequences, which are also input. How about using `projected` instead of `input`? 

Also, does [B, S, H, D] denote a fused QKV tensor, or separate Q, K, V tensors`?

[the remark this answers is not in the document]
Comment (samadejacobs) [reply blogs/ulysses-offload/README.md]: Separate

**Review by tjruwase — APPROVED** [review]
